File: Lynx-IOS/Lynx-IOS/Networking/SSEClient.swift
import Foundation

enum SSEError: Error {
    case httpError(Int)
}

/// Streams the SSE response body of POST /v1/conversations/:id/messages
/// and yields decoded `SSEEvent`s as they arrive.
enum SSEClient {
    static func stream(
        conversationId: String,
        content: String,
        accessToken: String
    ) -> AsyncThrowingStream<SSEEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    var request = URLRequest(
                        url: AppConfig.apiBaseURL.appending(path: "/v1/conversations/\(conversationId)/messages")
                    )
                    request.httpMethod = "POST"
                    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
                    request.httpBody = try JSONEncoder().encode(SendMessageBody(content: content))

                    let (bytes, response) = try await URLSession.shared.bytes(for: request)
                    if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
                        throw SSEError.httpError(http.statusCode)
                    }

                    var parser = SSEParser()
                    var lineBuffer = Data()
                    for try await byte in bytes {
                        lineBuffer.append(byte)
                        // Decode eagerly in UTF-8-safe chunks by flushing on newlines,
                        // matching the web client's TextDecoder(stream:true) behavior.
                        if byte == UInt8(ascii: "\n"), let chunk = String(data: lineBuffer, encoding: .utf8) {
                            for event in parser.feed(chunk) {
                                continuation.yield(event)
                            }
                            lineBuffer.removeAll(keepingCapacity: true)
                        }
                    }
                    if !lineBuffer.isEmpty, let chunk = String(data: lineBuffer, encoding: .utf8) {
                        for event in parser.feed(chunk) {
                            continuation.yield(event)
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}

File: Lynx-Andrioid/app/src/main/java/com/anantha/lynx/network/SseClient.kt
package com.anantha.lynx.network

import com.anantha.lynx.model.SendMessageBody
import com.anantha.lynx.model.SseEvent
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.Dispatchers
import kotlinx.serialization.json.Json
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit

class SseException(val code: Int) : Exception("SSE request failed with HTTP $code")

/**
 * Streams the SSE response body of POST /v1/conversations/:id/messages and
 * yields decoded SseEvents as they arrive. Ported from
 * Lynx-IOS/Lynx-IOS/Networking/SSEClient.swift.
 */
object SseClient {
    // A dedicated client with no read timeout — the default would kill a
    // long-lived streaming response.
    private val streamingHttpClient = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    private val json = Json { ignoreUnknownKeys = true }

    fun stream(conversationId: String, content: String, accessToken: String): Flow<SseEvent> = callbackFlow {
        val body = json.encodeToString(SendMessageBody.serializer(), SendMessageBody(content))
            .toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(
                AppConfig.API_BASE_URL.toHttpUrl().newBuilder()
                    .addPathSegments("v1/conversations/$conversationId/messages")
                    .build()
            )
            .addHeader("Authorization", "Bearer $accessToken")
            .post(body)
            .build()

        val call = streamingHttpClient.newCall(request)

        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                close(e)
            }

            override fun onResponse(call: Call, response: Response) {
                response.use { resp ->
                    if (!resp.isSuccessful) {
                        close(SseException(resp.code))
                        return
                    }

                    val parser = SseParser()
                    val source = resp.body.source()
                    try {
                        while (!source.exhausted()) {
                            val line = source.readUtf8Line() ?: break
                            for (event in parser.feed("$line\n")) {
                                trySend(event)
                            }
                        }
                        close()
                    } catch (e: IOException) {
                        close(e)
                    }
                }
            }
        })

        awaitClose { call.cancel() }
    }.flowOn(Dispatchers.IO)
}
